### ISSUE TITLE
Support CSC matrix

### ISSUE THREAD

**Issue by kaizhang** [kaizhang/anndata-rs#5]
#4

**Comment by zqfang** [comment]
Hi Kai,


I have critical issues with extracting data values from the `X` slot. 

the `select` or `slice` method is quite slow. 

eg. to extract a gene value (a column)  from dataset ( ~40k cells with 30k genes) in the `X` slot, it took about 20 seconds to get the expression value.

I've tested many times in the Rust side, it is just very slow. should CSR matrix to CSC matrix work better ? 

Would you give me some advice to get a blaze-fast version ?

The code I use are following
```rust
/// extract a column from adata.X
fn get_gene_exp(idx: usize) -> ArrayData {
    let msg = format!("Get gene with index: {}", idx);
    println!("{}", msg);
    let adata = ANNDATA_PROVIDER.get().unwrap();
    let idxx = vec![idx];
    let gixx = s![.., idxx];
    let start = Instant::now();
    let mut gex0 = adata.get_x().inner();
    // gex0.enable_cache();
    let end1 = Instant::now();
    let gex = gex0.select::<ArrayData, _>(gixx.as_ref()).unwrap();
    let end2 = Instant::now();
    // let gex = adata.x().slice::<ArrayData, _>(gixx).unwrap().unwrap();// CSRMatrix
    println!("get x time: {:.2?}", end1.duration_since(start));
    println!("get gene time: {:.2?}", end2.duration_since(end1));
    return gex;
}
```

**Comment by kaizhang** [comment]
This is expected, as in general selecting columns from a CSR matrix is slow. When it comes to on-disk representation, the situation becomes worse. I've run a few benchmarks and it shows that reading a portion of the matrix from the disk is no more performant than reading the whole matrix and then performing subsetting. So right now, any column indexing will force the entire matrix to be read. To speed this up, you can try:

1. Make batch query.
2. Enable cache. The 1st query will still be slow. But subsequent indexing should be fast.

**Comment by zqfang** [comment]
I've tried  `gex0.enable_cache()`, but it seems not help

**Comment by zqfang** [comment]
for a small dataset, it's fine.

But for larger datasets, it's just too slow to make everything work.

**Comment by kaizhang** [comment]
> for a small dataset, it's fine.
> 
> But for larger datasets, it's just too slow to make everything work.

Using CSC matrix for column indexing should make it much faster. Using other backends may also improve the speed. Once there is a mature Rust implementation of the Zarr format, I will switch from HDF5 to Zarr format.

Adding CSC support is fairly straightforward. But I'm working on something else right now. Let me know if you want to give it a try.

**Comment by zqfang** [comment]
I would wait some time.

**Comment by zqfang** [comment]
Hi Kai,

Just let you know. when I used a released binary, the indexing speed has 5X faster, `enable_cache` makes it to 10X. (~2s indexing time for my data)

**Comment by zqfang** [comment]
Hi @kaizhang , how is the current status of CSC matrix support? Could you update the code base if you've already implement something ?  Thanks

**Comment by kaizhang** [comment]
Hi @zqfang , it isn't that hard but I'm got caught by other projects. Do you want to give it a try?

**Comment by zqfang** [comment]
I can implement it recently. So i will just `copy` the code of `CsrMatrix`  and change to `CscMatrix` first

**Comment by zqfang** [comment]
Hi @kaizhang , Would you do a code review for the CSC implementation support ? I'll make a PR if you have time

**Comment by kaizhang** [comment]
Yes, please feel free to send a pull request. I will merge it once I have more time next month. Thank you!

**Comment by kaizhang** [comment]
fixed in b5adca519cf42ff5671ef2e9716b5fc1d99b43fa
